Core/Src/internalResistance.c

```diff
@@ -0,0 +1,221 @@
+/* ==================================================================== */
+/* ============================= INCLUDES ============================= */
+/* ==================================================================== */
+
+#include "internalResistance.h"
+#include "math.h"
+
+/* ==================================================================== */
+/* ========================= LOCAL VARIABLES ========================== */
+/* ==================================================================== */
+
+// Discrete buffers hold consecutive sensor readings from the bms
+static float currentDiscreteBuffer[DISCRETE_CURRENT_BUFFER_SIZE];
+static float voltageDiscreteBuffer[NUM_BMBS_PER_PACK][NUM_BRICKS_PER_BMB][DISCRETE_VOLTAGE_BUFFER_SIZE];
+
+// Discrete buffer pointers hold current index of discrete buffers
+static uint32_t discreteCurrentBufferIndex = 0;
+static uint32_t discreteVoltageBufferIndex = 0;
+
+// Data good indicates that all sensor readings in discrete buffers are good
+static bool currentDataGood = true;
+static bool voltageDataGood[NUM_BMBS_PER_PACK][NUM_BRICKS_PER_BMB] = {true};
+
+// Data ready indicates that the discrete buffer has filled
+static bool currentDataReady = false;
+static bool voltageDataReady = false;
+
+/* ==================================================================== */
+/* ==================== LOCAL FUNCTION DEFINITIONS ==================== */
+/* ==================================================================== */
+
+static void calculateInternalResistance(Bms_S* bms)
+{
+    // Circular average buffers hold the average of the discrete buffers, updated each time both discrete buffers fill
+    // Current buffer is initialized to a placeholder value to indicate data has not been placed in yet
+    static float currentAvgBuffer[AVERAGE_BUFFER_SIZE] = {[0 ... AVERAGE_BUFFER_SIZE-1] = -1000.0f};
+    static float voltageAvgBuffer[NUM_BMBS_PER_PACK][NUM_BRICKS_PER_BMB][AVERAGE_BUFFER_SIZE];
+    static uint32_t avgBufferIndex = 0;
+
+    /*  An average can only be added to the buffer if the current sensor
+        did not return any faulty data to the discrete buffer   */
+    if(currentDataGood)
```

Comment:
I think this function should be split in two. We should have one function that updates the averageBuffer and then one function that calculates the internal resistance. You'll have to move the currentAvgBuffer and voltageAvgBuffer out of the  function for this.